### ISSUE TITLE
Add Action properties

### ISSUE THREAD

**Issue by asherber** [CollaboratingPlatypus/PetaPoco#489]
Working through some of the issues in #488, I wanted to override `Database.OnExecutingCommand()`. This meant defining my own `Database` descendant and then implementing in that descendant any of the `Database` constructors I might need.

It occurred to me that customization would be easier if `Database` provided an `Action` property for each of the `OnXxx()` methods. That would give the user an easier way of customizing the behavior of a `Database` instance without having to define a whole new class.

For example:

```csharp
public class Database: IDatabase
{
    public Action<IDbCommand> ExecutingCommand { get; set; }
    
    public virtual void OnExecutingCommand(IDbCommand cmd)
    {
        ExecutingCommand?.Invoke(cmd);
    }
}

using (var db = new Database())
{
    db.ExecutingCommand = cmd => cmd.Parameters.Add(new SqlParameter());
}
```

This wouldn't be a breaking change, because anybody with a `Database` descendant that overrides the default implementation of `OnExecutingCommand()` will still get their override. 

Is this desirable? I can prepare a PR. (As a side note, I think the `OnXxx()` methods should probably be `protected` instead of `public`, although this would technically be a breaking change.)

**Comment by pleb** [comment]
@asherber yeah the current hooks are, well, err, not the best. On the roadmap, there are plans to make them more flexible. 

I'm not a huge fan of .net events, but they do allow multiple subs. They do, however, tend to become a source of leaks as people forget to unsubscribe. Just did a quick google search and found this (https://docs.microsoft.com/en-us/dotnet/framework/wpf/advanced/weak-event-patterns)

Ideally, one should be able to add hooks at a global level, like when defining/using the fluent configuration. And, at a per instance level.

I guess it comes down to

Delegates(plane or events)\Observer (IObservable<T> etc)\Custom

🤔

**Comment by asherber** [comment]
Personally, I would vote for delegates for this scenario. My suggestion above is lightweight, just passing in the `cmd` object, but it would be easy to implement a regular `EventHandler` pattern as well. `IObservable` feels like overkill, and to my mind it's not the right pattern for what this situation needs.

I like the idea of making the events accessible in the fluent API. Something like `config.WithExecutingCommand(cmd => cmd.Foo());`

(I have some thoughts about the article you posted as well, but I don't want to sidetrack this discussion. TL;DR: I don't think it's an issue here.)
